File: Solutions/p1_B.py
#!/usr/bin/env python3
# RIGHT CODE
import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from math import pow,sqrt
import transformations as tf
import math

class go_to_goal:
    def __init__(self):
        # Intialize the the turtle_controller node
        rospy.init_node('turtle_controller',anonymous=True)   

        # Create Publisher and Subscriber object
        pub_topic="/turtle1/cmd_vel"                          
        sub_topic="/turtle1/pose"                            

        self.pub = rospy.Publisher(pub_topic, Twist, queue_size=10)
        self.sub = rospy.Subscriber(sub_topic, Pose, self.update_pose)

        # Create pose object
        self.pose=Pose()                                       
        self.rate =rospy.Rate(50)
        
    def update_pose(self, data):
        # Update the pose from the received data DATA RECEIVED CORRECTLY
        # print("Received pose data:", data)
        self.pose.x = data.x
        self.pose.y = data.y
        self.pose.theta = data.theta  
        # print("Updated self.pose:", self.pose)
        
    
    def euclidean_distance(self,goal_pose):
        # Calculate the euclidean distance between desired and current position
        return sqrt(pow((goal_pose.x - self.pose.x),2) + pow((goal_pose.y - self.pose.y),2))
    
    def linear_vel(self,goal_pose,kp):
        # Calculate linear velocity based on euclidean distance. Kp is gain parameter
        return kp*(self.euclidean_distance(goal_pose))
    
    def heading_dirn(self,goal_pose):
        # Calculate the heading diretion
        delta_y = goal_pose.y - self.pose.y
        delta_x = goal_pose.x - self.pose.x
        return math.atan2(delta_y, delta_x)
    
    def calculate_angle_difference(self,angle2, angle1): 
        # Create quaternion from angle1 and angle2
        q1 = tf.transformations.quaternion_from_euler(0, 0,angle1)
        q2 = tf.transformations.quaternion_from_euler(0, 0,angle2)
        
        # Calculate quaternion difference
        q_diff = tf.transformations.quaternion_multiply(q2, tf.transformations.quaternion_inverse(q1))
        
        # Convert quaternion difference back to angles
        roll, pitch, yaw = tf.transformations.euler_from_quaternion(q_diff)
        angle_diff = yaw
        return angle_diff
    
    def angular_vel(self,goal_pose,kw):
        # Calculate the angular velocities
        return kw*self.calculate_angle_difference(self.heading_dirn(goal_pose),self.pose.theta)

    def going_to_goal(self,x,y):
        goal_pose=Pose()

        goal_pose.x= x #1.42
        goal_pose.y= y  #9
        
        # Tolerance
        error=0.001

        # Create Twist Object
        vel_msg=Twist() 
 
        # Move until the end   
        rospy.sleep(1.0)                                   
        while self.euclidean_distance(goal_pose)>=error:
            vel_msg.linear.x = self.linear_vel(goal_pose,1.5)
            vel_msg.linear.y = 0
            vel_msg.linear.z = 0

            vel_msg.angular.x=0
            vel_msg.angular.y=0
            vel_msg.angular.z=self.angular_vel(goal_pose,10)
            self.pub.publish(vel_msg)
            self.rate.sleep()

        # Stop the robot
        vel_msg.linear.x=0
        vel_msg.angular.z=0
        self.pub.publish(vel_msg)                               # Publishing velocity commands

        rospy.sleep(1.0)
        print(f"Euclidean distance Error: {self.euclidean_distance(goal_pose):.4f}, (X, Y) = ({self.pose.x:.4f}, {self.pose.y:.4f})")

    def orient_myway(self,goal_theta):
        # This function orients the turtle in desired orientation which was not possible with going_to_goal function
        goal_pose=Pose()
        goal_pose.theta=goal_theta
        vel_msg=Twist()
        while goal_pose.theta - self.pose.theta >=0.001:
            vel_msg.angular.z=3*(goal_pose.theta-self.pose.theta)
            self.pub.publish(vel_msg)
            self.rate.sleep()

        vel_msg.linear.x=0
        vel_msg.angular.z=0
        self.pub.publish(vel_msg)                               # Publishing velocity commands
        print(f"Heading Error: {goal_pose.theta - self.pose.theta:.4f}")

if __name__=="__main__":
    try:

        x=float(input("Enter the x-coordinate: "))
        y=float(input("Enter the y-coordinate: "))
        turtle=go_to_goal()
        print("************************************************************")
        turtle.going_to_goal(x,y)
        print("Reached first goal. \n", flush=True)
        print("************************************************************")
        # turtle.going_to_goal(y,x)
        # print("Reached second goal. \n", flush=True)
        # print("************************************************************")
        # turtle.going_to_goal(5.45,5.45)
        # print("Reached third goal. \n", flush=True)
        # print("************************************************************")
        # turtle.orient_myway(1.57)
        # print("Oriented as per my way \n",flush=True)
    except rospy.ROSInterruptException:
        pass       
        

File: Solutions/p1_B.cpp
#include <iostream>
#include <cstdlib>
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "turtlesim/Pose.h"
#include "cmath"
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <tf2/convert.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

class TurtleController {

private:
    ros::NodeHandle* nh;            // Pointer to ros::NodeHandle object for creating ROS nodes
    ros::Publisher pub;             // Create a publisher object
    ros::Subscriber sub;            // Create a subscriber object
    std::string pub_topic;          // String variable to store name of topic on which we publish
    std::string sub_topic;          // String variable to store name of topic to which we subscribe
    turtlesim::Pose pose;           // Create pose object
    ros::Rate* rate;                // pointer to rate object to control rate of ROS node's main loop iterations (int main)
    double kp,kw;

public:
    TurtleController(int argc,char** argv) {
        // Initialize ROS node
        ros::init(argc, argv, "turtle_controller_cpp");
        nh = new ros::NodeHandle();

        // Publisher and subscriber topics
        pub_topic = "/turtle1/cmd_vel";
        sub_topic = "/turtle1/pose";

        // Create publisher and subscriber
        pub = nh->advertise<geometry_msgs::Twist>(pub_topic, 10);
        sub = nh->subscribe(sub_topic, 10, &TurtleController::updatePose, this);

        // Initialize pose object and rate
        pose = turtlesim::Pose();
        rate = new ros::Rate(100);
    }

    // Callback to update the pose
    void updatePose(const turtlesim::Pose::ConstPtr& msg) {
        // Update pose object using msg parameter
        pose.x = msg->x;
        pose.y = msg->y;
        pose.theta =msg->theta;
        
    }

    //  Calculate the euclidean distance between desired and current position
    double euclideanDistance(const turtlesim::Pose& goalPose){
        return std::sqrt(std::pow(goalPose.x - pose.x,2) + std::pow(goalPose.y - pose.y,2));
    }

    // Calculate linear velocity based on euclidean distance. Kp is gain parameter
    double linearVel(const turtlesim::Pose& goalPose,double kp){
        return kp*(euclideanDistance(goalPose));
    }

    // Calculate the heading direction
    double headingdirn(const turtlesim::Pose& goalPose){
        return std::atan2(goalPose.y - pose.y,goalPose.x - pose.x);
    }

    // Create quaternions from angle1 and angle2
    double calculateAngleDifference(double angle2, double angle1) {
        tf2::Quaternion q1;
        q1.setRPY(0, 0, angle1);
        
        tf2::Quaternion q2;
        q2.setRPY(0, 0, angle2);
        
        // Calculate quaternion difference
        tf2::Quaternion q_diff = q2 * q1.inverse();
        
        // Convert quaternion difference back to angles
        double roll, pitch, yaw;
        tf2::Matrix3x3(q_diff).getRPY(roll, pitch, yaw);

        double angle_diff = yaw;
        
    return angle_diff;
    }

    // Calculates the angular velocities
    double angularVel(const turtlesim::Pose& goalPose,double kw){
        return kw*(calculateAngleDifference(headingdirn(goalPose),pose.theta));
    }

    // Move until we are there
    void going_to_goal(double x,double y){
        turtlesim::Pose goalPose;

        goalPose.x=x;
        goalPose.y=y;

        pose.x=5.5444;
        pose.y=5.5444;
        
        double error = 0.05;

        //Using geometry_msgs::Twist for Twist Object
        geometry_msgs::Twist velMsg; 
        while (euclideanDistance(goalPose)>=0.01){
            // std::cout<<"x"<<pose.x<<std::endl;
            // std::cout<<"y"<<pose.y<<std::endl;
            velMsg.linear.x = linearVel(goalPose,1.5);
            velMsg.linear.y = 0;
            velMsg.linear.z = 0;

            velMsg.angular.x = 0;
            velMsg.angular.y = 0;
            velMsg.angular.z = angularVel(goalPose,6.0);

            pub.publish(velMsg);
            ros::spinOnce();  // process callbacks for subscribers
            rate->sleep();
            pose=pose;  // update the pose 
        }

        // stop the robot
        velMsg.linear.x=0;
        velMsg.angular.z=0;
        pub.publish(velMsg);
        
        ROS_INFO_STREAM("Euclidean distance Error: " << euclideanDistance(goalPose) << ", (X,Y) = (" << pose.x <<","<< pose.y << ")");
    
    }
    void orient_myway(double goal_theta){
        //This function orients the turtle in desired orientation which was not possible with going_to_goal function 
        turtlesim::Pose goalPose;
        goalPose.theta = goal_theta;

        geometry_msgs::Twist velMsg;
        while(goalPose.theta - pose.theta >=0.1){
            velMsg.angular.z = 1*(goalPose.theta - pose.theta);
            pub.publish(velMsg);
            ros::spinOnce();   /// a very important command
            rate->sleep();
            pose=pose;
        }
        // stop the robot
        velMsg.linear.x=0;
        velMsg.angular.z=0;
        pub.publish(velMsg);
        
        ROS_INFO_STREAM("Heading Error:" <<goalPose.theta - pose.theta);   
    }

};

int main(int argc, char** argv) {

        float x,y;

        std::cout << "Enter the x-coordinate in range (0, 11): ";
        std::cin >> x;
        std::cout << "Enter the y-coordinate in range (0, 11): ";
        std::cin >> y;
        ros::init(argc,argv,"turtle_controller_cpp");
    

    try{
        ros::NodeHandle nh;
        TurtleController turtle(argc,argv);
        std::cout << "************************************************************" << std::endl;
        turtle.going_to_goal(x,y);
        std::cout<< " Reached first goal."<<std::endl;
        std::cout << "************************************************************" << std::endl;
        // turtle.going_to_goal(5.45,9.45);
        // std::cout<< " Reached second goal."<<std::endl;
        // std::cout << "************************************************************" << std::endl;
        // turtle.going_to_goal(5.45,5.45);
        // std::cout<< " Reached third goal."<<std::endl;
        // std::cout << "************************************************************" << std::endl;
        // turtle.orient_myway(1.57);
        // std::cout<< " Oriented as per my way"<<std::endl;
        // std::cout << "************************************************************" << std::endl;

        ros::spin();
        return 0;
    } catch (const ros::Exception& e ){
        ROS_INFO("ROS Interrupt Exception: %s", e.what());

    }
    return 0;
}




